README.rst
mongodict - MongoDB-backed Python dict-like interface
=====================================================

So you are storing some key-values in a ``dict`` but your data became huge than
your memory or you want to persist it on the disk? Then `mongodict
<https://github.com/turicas/mongodict>`_ is for
you!

As it uses `MongoDB <http://mongodb.org/>`_ to store the data, you get all cool
`MongoDB <http://mongodb.org/>`_ things, like shardings and replicas. It uses
the `pickle` module available on Python standard library to
serialize/deserialize data and store everything as `bson.Binary` in MongoDB.
You can also provide another codec (serializer/deserializer).

`mongodict <https://github.com/turicas/mongodict>`_ is tested under
`Python 2.7.5 <http://www.python.org/getit/releases/2.7/>`_ and
`Python 3.3.2 <http://www.python.org/getit/releases/3.2/>`_.


Installation
------------

As simple as::

    pip install mongodict


Usage
-----

As it uses
`collections.MutableMapping <http://docs.python.org/library/collections.html#collections.MutableMapping>`_
as its base, you just need to change the line which creates your ``dict``.
For instace, just replace::

    >>> my_dict = {}

with::

    >>> from mongodict import MongoDict
    >>> my_dict = MongoDict(host='localhost', port=27017, database='my_dict',
                            collection='store')

and then use it like a normal ``dict``::

    >>> my_dict['python'] = 'rules'
    >>> print my_dict['python']
    rules
    >>> del my_dict['python']
    >>> print my_dict['python']
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
      File "mongodict.py", line 82, in __getitem__
        raise KeyError(key)
    KeyError: u'python'
    >>> my_dict['spam'] = 'eggs'
    >>> my_dict['ham'] = 'damn'
    >>> for key, value in my_dict.items():
    ...    print '{} = {}'.format(key, value)
    ...
    spam = eggs
    ham = damn

If you want to use another codec, you should pass serialize and deserialize
functions to the class during the initialization. For example, to use JSON::

    >>> import json
    >>> json_dict = MongoDict(host='localhost', port=27017,
                              database='json_dict', collection='store',
                              codec=(json.dumps, json.loads))
    >>> # use json_dict as usual

Enjoy! :-)

.. NOTE::
   There is no kind of in-memory cache, so all key lookups will be translated
   in a `MongoDB <http://mongodb.org/>`_ query but as
   `MongoDB <http://mongodb.org/>`_'s server put everything it can in memory,
   probably it'll not be a problem (if your working set is always entire in
   memory).


Authentication
--------------

If you want to use MongoDB's authentication to the database ``MongoDict`` is
connecting to, you just need to provide an ``auth`` parameter, as in this
example::


    from mongodict import MongoDict


    my_dict = MongoDict(host='localhost', port=27017, database='mydb',
                        collection='mongodict',
                        auth=('my username', 'my password'))



Why not Redis?
--------------

`Redis <http://redis.io/>`_ is "remote directory server" - it's a great piece
of software and can do the job if all your data fit on memory. By other side,
`MongoDB <http://mongodb.org/>`_ already have mature sharding and replica set
features. So, if you need to store lots of key-value pairs that don't fit on
memory, `mongodict <https://github.com/turicas/mongodict>`_ can solve your
problem.

.. NOTE::
   `mongodict <https://github.com/turicas/mongodict>`_ does not have the same
   API other key-value software have
   (like `memcached <http://memcached.org/>`_). Some features are missing to
   compete directly with these kind of software (by now we have only the
   ``dict``-like behaviour), but I have plans to add it soon.


Contributing
------------

- This software uses `Semantic Versioning <http://semver.org/>`_.
- This software uses
  `Test-Driven Development <http://en.wikipedia.org/wiki/Test-driven_development>`_.

You can run the tests either with or without
`tox <http://tox.readthedocs.org/en/latest/index.html>`_.

Without tox
~~~~~~~~~~~

This is the simplest approach: you'll test only for one Python version. To do
it, just execute::

    mkvirtualenv --no-site-packages mongodict-without-tox
    pip install -r requirements/develop.txt
    make test


With tox
~~~~~~~~

With `tox <http://tox.readthedocs.org/en/latest/index.html>`_ you can test for
more than one Python version (currently for 2.7 and 3.2). You just need to
create a virtualenv, install and run it::

    mkvirtualenv --no-site-packages tox-for-mongodict
    pip install tox
    tox

`tox <http://tox.readthedocs.org/en/latest/index.html>`_ will create one
virtualenv for each Python version, install requirements and then run the tests
for each of them. Note that you need the python binaries available in your
system (2.7 and 3.2) to run the tests.


Author
------

This software was written and is maintained by
`Álvaro Justen (aka Turicas) <https://github.com/turicas>`_.
Please contact me at ``alvarojusten`` *at* ``gmail`` *dot* ``com``.


License
-------

It's licensed under `GPL version 3 <https://www.gnu.org/licenses/gpl-3.0.html>`_.


CHANGELOG.rst
Change Log
==========

Version 0.3.1
-------------

This version BREAKS DATA COMPATIBILITY with 0.3.0. Please see below:

- Option to specify index type: key-only or key-and-value. If you specify "key
  and value" you can't have documents (key + value) greater than 16MB. Default
  option moved to key-only (fixes `issue #14
  <https://github.com/turicas/mongodict/issues/14>`_);
- Replace ``value`` with ``v`` as the key inside Mongo document to save space.
  Note that you will need to migrate your data! A little script is provided.


Version 0.3.0
-------------

- Added a "codec" concept, so user can customized how value data is saved.
  `Pickle is the default codec <https://github.com/turicas/mongodict/issues/12>`_;
- Added a
  `migration script <https://github.com/turicas/mongodict/blob/master/migrate_data.py>`_
  to migrate data from mongodict <= 0.2.1;
- `Using basic mapping (dict) interface tests from python source
  <https://github.com/turicas/mongodict/issues/13>`_;
- Updated `tox <https://pypi.python.org/pypi/tox>`_ config to use Python 2.7.5
  and Python 3.3.2 to run tests.


Version 0.2.1
-------------

- [Bug] Fix utf8 encoding problem when using ``bson.Binary`` (#11)
- [Bug] Fix 16MB-limit problem when calling ``MongoDict.keys`` (#10)
- [Bug] ``MongoDict.clear()`` do not remove collection indexes anymore
- Raise *instance* of ``KeyError`` instead of class
- Add ``auth`` parameter to ``MongoDict`` (#8)
- Add information about tox in README
- Fixes setup.py to work with python 2.6
- Use ``collection.count()`` instead of ``cursor.count()``


Version 0.2.0
-------------

- Add Python 3.2 support (thanks, @jucacrispim)
- Optimize all queries to hit only the index
- Add docstrings to entire module
- Add ``mongodict.__all__``
- Add ``MongoDict.__del__``
- Use tox to run tests


Version 0.1.1
-------------

- Set ``safe=True`` by default (slower)
- Did some optimizations on ``__getitem__`` and  ``__delitem__``
- Added method ``clear``, an optimized way to delete all keys
- An ``UnicodeDecodeError`` is raised if one of the strings
  (key or value) is not ``unicode`` or ``str`` encoded with ``utf-8`` codec
- Replaced ``insert`` with ``update`` - it was generating an exception when
  updating (already existent) keys
- Add benchmark (Python dict vs Redis vs mongodict)


Version 0.1.0
-------------

- First version released
- Basic ``dict``-like behaviour


test_mongodict.py
# coding: utf-8

import sys
import unittest

from collections import MutableMapping

import pymongo

from bson import Binary
from mongodict import MongoDict


class TestMongoDict(unittest.TestCase):
    def setUp(self):
        self.config = {'host': 'localhost', 'port': 27017,
                       'database': 'mongodict', 'collection': 'main'}
        self.connection = pymongo.Connection(host=self.config['host'],
                port=self.config['port'], safe=True)
        self.db = self.connection[self.config['database']]
        self.collection = self.db[self.config['collection']]

    def tearDown(self):
        self.connection.drop_database(self.db)

    def test_should_be_a_MutableMapping(self):
        my_dict = MongoDict()
        self.assertTrue(isinstance(my_dict, MutableMapping))
        expected_methods = ['setitem', 'getitem', 'delitem', 'iter', 'len']
        actual_methods = dir(my_dict)
        for method in expected_methods:
            self.assertIn('__{}__'.format(method), actual_methods)

    def test_set_item(self):
        my_dict = MongoDict(**self.config)
        my_dict['python'] = 'rules'
        results = list(self.collection.find())
        self.assertEqual(len(results), 1)
        self.assertEqual(results[0]['_id'], 'python')
        self.assertEqual(results[0]['value'], 'rules')

    def test_default_items(self):
        my_dict = MongoDict(default={'answer': 42, 'spam': 'ham'},
                            **self.config)
        results = list(self.collection.find())
        self.assertEqual(len(results), 2)
        self.assertEqual(results[0]['_id'], 'answer')
        self.assertEqual(results[0]['value'], 42)
        self.assertEqual(results[1]['_id'], 'spam')
        self.assertEqual(results[1]['value'], 'ham')

    def test_get_item(self):
        self.collection.insert({'_id': 'testing', 'value': 123})
        self.collection.insert({'_id': 'bla bla bla', 'value': 3.14})
        my_dict = MongoDict(**self.config)
        self.assertEqual(my_dict['testing'], 123)
        self.assertEqual(my_dict['bla bla bla'], 3.14)
        with self.assertRaises(KeyError):
            temp = my_dict['non ecxiste']

    def test_del_item(self):
        self.collection.insert({'_id': 'testing', 'value': 123})
        self.collection.insert({'_id': 'bla bla bla', 'value': 3.14})
        my_dict = MongoDict(**self.config)
        del my_dict['testing']
        results = list(self.collection.find())
        self.assertEqual(results, [{'_id': 'bla bla bla', 'value': 3.14}])
        with self.assertRaises(KeyError):
            del my_dict['non ecxiste']

    def test_len(self):
        for counter in range(1000):
            self.collection.insert({'_id': 'test-' + str(counter),
                                    'value': counter})
        my_dict = MongoDict(**self.config)
        self.assertEqual(self.collection.find().count(), len(my_dict))

    def test_iter(self):
        for counter in range(10):
            self.collection.insert({'_id': 'test-' + str(counter),
                                    'value': counter})
        my_dict = MongoDict(**self.config)
        keys = []
        for key in my_dict:
            keys.append(key)
        self.assertEqual(len(keys), 10)
        expected_keys = ['test-' + str(counter) for counter in range(10)]
        self.assertEqual(set(keys), set(expected_keys))
        self.assertEqual(set(my_dict.keys()), set(expected_keys))
        my_dict['python'] = 10
        expected_keys.append('python')
        self.assertEqual(set(my_dict.keys()), set(expected_keys))

        results = []
        for key, value in my_dict.items():
            results.append((key, value))
        values = [x[1] for x in results]
        expected_values = list(range(11))
        self.assertEqual(set(values), set(expected_values))
        self.assertEqual(set(my_dict.values()), set(expected_values))

    def test_in(self):
        self.collection.insert({'_id': 'testing', 'value': 123})
        my_dict = MongoDict(**self.config)
        self.assertIn('testing', my_dict)
        self.assertNotIn('python', my_dict)

    def test_clear(self):
        for counter in range(10):
            self.collection.insert({'_id': 'test-' + str(counter),
                                    'value': counter})
        my_dict = MongoDict(**self.config)
        index_count_before = len(self.collection.index_information())
        my_dict.clear()
        index_count_after = len(self.collection.index_information())
        self.assertEqual(self.collection.find().count(), 0)

        # MongoDict.clear() should not delete indexes
        self.assertEqual(index_count_before, index_count_after)

    def test_should_be_possible_to_assign_new_values_to_existing_keys(self):
        my_dict = MongoDict(**self.config)
        my_dict['python'] = 'rules'
        my_dict['python'] = 42
        self.assertNotEqual(my_dict['python'], 'rules')
        self.assertEqual(my_dict['python'], 42)


    def test_non_unicode_strings(self):
        my_dict = MongoDict(**self.config)
        if sys.version >= '3':
            return # pymongo on Python 3 does not have this problem

        string_1 = 'Álvaro Justen'.decode('utf8').encode('iso-8859-15')
        with self.assertRaises(pymongo.errors.InvalidStringData):
            my_dict[string_1] = 123
        with self.assertRaises(pymongo.errors.InvalidStringData):
            temp = my_dict[string_1]
        with self.assertRaises(pymongo.errors.InvalidStringData):
            my_dict['python'] = string_1
        with self.assertRaises(pymongo.errors.InvalidStringData):
            string_1 in my_dict
        with self.assertRaises(pymongo.errors.InvalidStringData):
            del my_dict[string_1]

    def test_deletion_of_MongoDict_object(self):
        config = self.config.copy()
        config['safe'] = False
        my_dict = MongoDict(**config)
        for i in range(1000):
            my_dict['testing_' + str(i)] = i
        del my_dict
        self.assertEqual(self.collection.find().count(), 1000)

    def test_keys_method_should_not_raises_exception_if_more_than_16MB(self):
        my_dict = MongoDict(**self.config)
        key_template = ('python-rules' * 100000) + '{}'
        key_byte_count = 0
        key_count = 0
        keys = set()
        while key_byte_count < 20 * 1024 * 1024: # 20MB > 16MB
            new_key = key_template.format(key_count)
            my_dict[new_key] = 'some value'
            key_byte_count += len(new_key)
            key_count += 1
            keys.add(new_key)
        dict_keys = my_dict.keys()
        self.assertEquals(len(keys), len(dict_keys))
        self.assertTrue(keys == set(dict_keys))
        # do not use assertEqual here! The content is too big

    def test_Binary(self):
        my_dict = MongoDict(**self.config)
        if sys.version < '3':
            invalid_utf8 = 'á'.decode('utf-8').encode('iso-8859-15')
        else:
            invalid_utf8 = 'á'.encode('iso-8859-15')
        binary_data = Binary(invalid_utf8)
        my_dict['test'] = binary_data
        result = my_dict['test']
        self.assertIn(type(result), (bytes, Binary))
        self.assertEqual(Binary(my_dict['test']), binary_data)


tests/test_mongodict.py
# coding: utf-8

import json
import pickle
import sys
import unittest

from collections import MutableMapping

import pymongo

from bson import Binary
from mongodict import MongoDict


if sys.version_info[0] < 3: # Python 2
    binary_type = str
    key_1 = 'Álvaro'.decode('utf-8')
    key_2 = 'Álvaro'
else:
    binary_type = bytes
    key_1 = 'Álvaro'
    key_2 = 'Álvaro'.encode('utf-8')


def encode(value):
    return pickle.dumps(value, protocol=pickle.HIGHEST_PROTOCOL)


def decode(value):
    return pickle.loads(value)


def extract_indexes(index_info):
    return [idx['key'] for idx in index_info.values()]


class TestMongoDict(unittest.TestCase):
    def setUp(self):
        self.config = {'host': 'localhost', 'port': 27017,
                       'database': 'mongodict', 'collection': 'main',}
        # as no codec is specified, it uses the default (pickle)
        self.connection = pymongo.Connection(host=self.config['host'],
                port=self.config['port'], safe=True)
        self.db = self.connection[self.config['database']]
        self.collection = self.db[self.config['collection']]

    def tearDown(self):
        self.connection.drop_database(self.db)

    def test_set_item_should_save_data_in_collection(self):
        my_dict = MongoDict(**self.config)
        my_dict['python'] = 'rules'
        results = list(self.collection.find())
        self.assertEqual(len(results), 1)
        self.assertEqual(results[0]['_id'], 'python')
        self.assertEqual(decode(results[0]['v']), 'rules')

    def test_get_item_should_retrieve_data_from_collection(self):
        self.collection.insert({'_id': 'testing',
                                'v': Binary(encode('123'))})
        self.collection.insert({'_id': 'bla bla bla',
                                'v': Binary(encode('3.14'))})
        my_dict = MongoDict(**self.config)
        self.assertEqual(my_dict['testing'], '123')
        self.assertEqual(my_dict['bla bla bla'], '3.14')
        with self.assertRaises(KeyError):
            temp = my_dict['non ecxiste']

    def test_del_item_should_delete_pair_in_the_collection(self):
        self.collection.insert({'_id': 'testing',
                                'v': Binary(encode('123'))})
        self.collection.insert({'_id': 'bla bla bla',
                                'v': Binary(encode('3.14'))})
        my_dict = MongoDict(**self.config)
        del my_dict['testing']
        results = list(self.collection.find())
        self.assertEqual(results[0]['_id'], 'bla bla bla')
        self.assertEqual(decode(results[0]['v']), '3.14')
        with self.assertRaises(KeyError):
            del my_dict['non ecxiste']

    def test_should_be_possible_to_assign_new_values_to_existing_keys(self):
        my_dict = MongoDict(**self.config)
        my_dict['python'] = 'rules'
        my_dict['python'] = '42'
        self.assertNotEqual(my_dict['python'], 'rules')
        self.assertEqual(my_dict['python'], '42')

    def test_deletion_of_MongoDict_object_should_sync_data_even_without_safe(self):
        config = self.config.copy()
        config['safe'] = False
        my_dict = MongoDict(**config)
        for i in range(1000):
            my_dict['testing_' + str(i)] = str(i)
        del my_dict
        self.assertEqual(self.collection.find().count(), 1000)

    def test_keys_method_should_not_raises_exception_if_more_than_16MB(self):
        '''Should not raise exception if sum of keys is greater 16MB

        Bug reported by @andrebco:
            <https://github.com/turicas/mongodict/issues/10>
        '''
        my_dict = MongoDict(**self.config)
        key_template = ('python-rules' * 100000) + '{}'
        key_byte_count = 0
        key_count = 0
        keys = set()
        while key_byte_count < 20 * 1024 * 1024: # 20MB > 16MB
            new_key = key_template.format(key_count)
            my_dict[new_key] = 'some value'
            key_byte_count += len(new_key)
            key_count += 1
            keys.add(new_key)
        dict_keys = my_dict.keys()
        self.assertEquals(len(keys), len(dict_keys))
        self.assertTrue(keys == set(dict_keys))
        # do not use assertEqual here because the content is too big and if it
        # fails, stdout will be flooded by irrelevant data

    def test_pickle_codec_should_return_same_objects(self):
        self.collection.drop()
        my_dict = MongoDict(**self.config)
        my_dict['int'] = 42
        my_dict['float'] = 3.14
        my_dict['string'] = 'python'
        my_dict[key_2] = [3, 4]
        my_dict[key_1] = {123: 456}
        # since key_1 and key_2 has the same information (one is unicode, other
        # is bytes), the contents are overwritten in MongoDict
        self.assertEqual(my_dict['int'], 42)
        self.assertEqual(my_dict['float'], 3.14)
        self.assertEqual(my_dict['string'], 'python')
        self.assertEqual(my_dict[key_1], {123: 456})

    def test_customized_codec(self):
        self.collection.drop()
        self.config['codec'] = (lambda x: json.dumps(x).encode('utf-8'),
                                json.loads)
        data = {'first': (1, 2), 'test': [3, 4], 'test2': {'...': 456}}
        expected = [('first', [1, 2]), ('test', [3, 4]),
                    ('test2', {'...': 456})]
        # JSON represents Python tuples as arrays and deserialized arrays as
        # lists, so our tuples turn lists

        self.config['default'] = data
        my_dict = MongoDict(**self.config)

        for result in self.collection.find():
            key = result['_id']
            value = json.loads(result['v'].decode('utf-8'))
            pair = (key, value)
            self.assertIn(pair, expected)
            expected.remove(pair)
            #we need to remove every pair in this list (instead of just adding
            #it to a dict and then comparing the whole dicts) because JSON
            #encodes tuples as arrays and decodes arrays as lists and we can't
            #have a list as a dict key since it is not hashable
        self.assertEqual(expected, [])

    def test_verify_if_index_is_created(self):
        self.config['index_type'] = 'invalid'
        with self.assertRaises(ValueError):
            MongoDict(**self.config)
        del(self.config['index_type'])

        self.collection.drop()
        my_dict = MongoDict(**self.config) # default 'index_type' = 'key'
        indexes = extract_indexes(self.collection.index_information())
        expected_indexes = [[('_id', 1)]]
        self.assertEqual(indexes, expected_indexes)

        self.collection.drop()
        self.config['index_type'] = 'key-value'
        self.config['collection'] = 'index_test'
        self.collection = self.db[self.config['collection']]
        other_dict = MongoDict(**self.config)
        indexes = extract_indexes(self.collection.index_information())
        expected_indexes = set([(('_id', 1), ), (('_id', 1), ('v', 1))])
        new_indexes = set()
        for index in indexes:
            new_indexes.add(tuple(index))
        self.assertEqual(new_indexes, expected_indexes)

    # TODO: test types of keys (str, unicode)?


benchmark.py
# coding: utf-8

import timeit


test_code = '''
import random
import hashlib
import redis
from mongodict import MongoDict


data = {}
for counter in range(int(random.random() * 10000)):
    my_hash = hashlib.sha512(str(counter)).hexdigest()
    key = 'key-{}-{}'.format(str(counter), my_hash)
    value = hashlib.sha512(key).hexdigest()
    data[key] = value

def test_redis():
    redis_dict = redis.StrictRedis(host='localhost', port=6379, db=0)
    for key in redis_dict.keys():
        redis_dict.delete(key)
    for key, value in data.iteritems():
        redis_dict.set(key, value)
        assert redis_dict.get(key) == value

def test_mongodict(safe=True):
    mongo_dict = MongoDict(safe=safe)
    mongo_dict.clear()
    for key, value in data.iteritems():
        mongo_dict[key] = value
        assert mongo_dict[key] == value

def test_pythondict():
    python_dict = {}
    python_dict.clear()
    for key, value in data.iteritems():
        python_dict[key] = value
        assert python_dict[key] == value
'''

def main():
    iterations = 10
    tests = {'Python dict': 'test_pythondict()',
             'Redis': 'test_redis()',
             'MongoDict(safe=False)': 'test_mongodict(safe=False)',
             'MongoDict(safe=True)': 'test_mongodict(safe=True)',
    }
    for test_name, function_call in tests.iteritems():
        print test_name, timeit.timeit(function_call, test_code,
                                       number=iterations)
    #TODO: test for other number and sizes of keys/values

if __name__ == '__main__':
    main()


migrate_data.py
#!/usr/bin/env python
# coding: utf-8

'''
This script does data migration of data from mongodict <= 0.2.1 to 0.3.0 and
from 0.3.0 to 0.3.1.
The first migration is needed since the new version uses another codec by
default (pickle instead of JSON/BSON) and the later one because the key name
changes on document (`value` to `v`).
'''

import datetime
import sys
import time

import mongodict
import pymongo


encode_nothing = lambda x: x
decode_nothing = lambda x: x
report_template = ('\r{:010d} keys / {:010d} ({:6.2f}%) migrated. '
                   'Duration: {}, ETA: {}')
REPORT_INTERVAL = 10 # in number of keys


def print_report(counter, total, start_time):
    percentual = 100 * (counter / float(total))
    duration = datetime.timedelta((time.time() - start_time) / (24 * 3600))
    duration = str(duration).split('.')[0]
    rate = counter / (time.time() - start_time)
    eta = datetime.timedelta(((total - counter) / rate) / (24 * 3600))
    eta = str(eta).split('.')[0]
    report = report_template.format(counter, total, percentual,
                                    duration, eta)
    sys.stdout.write(report)
    sys.stdout.flush()


def migrate_codec(config_old, config_new):
    '''Migrate data from mongodict <= 0.2.1 to 0.3.0
    `config_old` and `config_new` should be dictionaries with the keys
    regarding to MongoDB server:
        - `host`
        - `port`
        - `database`
        - `collection`
    '''
    assert mongodict.__version__ in [(0, 3, 0), (0, 3, 1)]

    connection = pymongo.Connection(host=config_old['host'],
                                    port=config_old['port'])
    database = connection[config_old['database']]
    collection = database[config_old['collection']]
    new_dict = mongodict.MongoDict(**config_new) # uses pickle codec by default
    total_pairs = collection.count()
    start_time = time.time()
    for counter, pair in enumerate(collection.find(), start=1):
        key, value = pair['_id'], pair['value']
        new_dict[key] = value
        if counter % REPORT_INTERVAL == 0:
            print_report(counter, total_pairs, start_time)
    print_report(counter, total_pairs, start_time)
    print('')

def migrate_key(config_old, config_new):
    '''Migrate data from mongodict == 0.3.0 to 0.3.1
    `config_old` and `config_new` should be dictionaries with the keys
    regarding to MongoDB server:
        - `host`
        - `port`
        - `database`
        - `collection`
    '''
    assert mongodict.__version__ == (0, 3, 1)

    connection = pymongo.Connection(host=config_old['host'],
                                    port=config_old['port'])
    database = connection[config_old['database']]
    collection = database[config_old['collection']]
    new_dict = mongodict.MongoDict(**config_new) # uses `v` as default key
    total_pairs = collection.count()
    start_time = time.time()
    for counter, pair in enumerate(collection.find(), start=1):
        key, value = pair['_id'], new_dict.decode_value(pair['value'])
        new_dict[key] = value
        if counter % REPORT_INTERVAL == 0:
            print_report(counter, total_pairs, start_time)
    print_report(counter, total_pairs, start_time)
    print('')


def parse_mongo_data(data_as_string):
    info = data_as_string.split(':')
    port, database, collection = info[1].split('/')
    return {'host': info[0], 'port': int(port), 'database': database,
            'collection': collection}


def main():
    if mongodict.__version__[1] <= 2:
        sys.stderr.write('You need mongodict >= 0.3.0 to run this script.\n')
        exit(1)

    if len(sys.argv) < 4:
        error_message = ('ERROR: usage: {} <old host:port/db/coll> '
                         '<new host:port/db/coll> <codec|key>\n').format(sys.argv[0])
        sys.stderr.write(error_message)
        exit(2)

    try:
        old_config = parse_mongo_data(sys.argv[1])
        new_config = parse_mongo_data(sys.argv[2])
    except IndexError:
        sys.stderr.write('Error parsing MongoDB server data. Please check.\n')
        exit(3)
    migration_type = sys.argv[3]
    if migration_type not in ('codec', 'key'):
        sys.stderr.write('Invalid migration type: {}'.format(migration_type))
        exit(4)

    print('Starting migration...')
    if migration_type == 'codec':
        migrate_codec(old_config, new_config)
    elif migration_type == 'key':
        migrate_key(old_config, new_config)


if __name__ == '__main__':
    main()


mongodict.py
# coding: utf-8

''' MongoDB-backed Python ``dict``-like interface

So you are storing some key-values in a ``dict`` but your data became huge than
your memory or you want to persist it on the disk? Then `mongodict
<https://github.com/turicas/mongodict>`_ is for
you!

As it uses `MongoDB <http://mongodb.org/>`_ to store the data, you get all cool
`MongoDB <http://mongodb.org/>`_ things, like shardings and replicas.

`mongodict <https://github.com/turicas/mongodict>`_ is supported under
`Python 2.7 <http://www.python.org/getit/releases/2.7/>`_ and
`Python 3.2 <http://www.python.org/getit/releases/3.2/>`_.
'''

import sys
try:
    import cPickle as pickle
except ImportError:
    import pickle

from collections import MutableMapping

import pymongo

from bson import Binary


__version__ = (0, 3, 1)
__all__ = ['MongoDict']
INDEX_KEY = [('_id', 1)]
INDEX_KEY_VALUE = [('_id', 1), ('v', 1)]

if sys.version_info[0] == 2:
    binary_type = str
else:
    binary_type = bytes

def pickle_dumps(value):
    return pickle.dumps(value, protocol=pickle.HIGHEST_PROTOCOL)

class MongoDict(MutableMapping):
    ''' ``dict``-like interface for storing data in MongoDB '''


    def __init__(self, host='localhost', port=27017, database='mongodict',
                 collection='main', codec=(pickle_dumps, pickle.loads),
                 safe=True, auth=None, default=None, index_type='key'):
        ''' MongoDB-backed Python ``dict``-like interface

        Create a new MongoDB connection.
        `auth` must be (login, password)'''
        super(MongoDict, self).__init__()
        self._connection = pymongo.Connection(host=host, port=port, safe=safe)
        self._safe = safe
        self._db = self._connection[database]
        if auth is not None:  # TODO: test auth
            if not self._db.authenticate(*auth):
                raise ValueError('Cannot authenticate to MongoDB server.')
        self._collection = self._db[collection]
        if index_type == 'key':
            self._index = INDEX_KEY
        elif index_type == 'key-value':
            self._index = INDEX_KEY_VALUE
        else:
            raise ValueError(u'Error: unknown `index_type`')
        self._collection.ensure_index(self._index)
        self.encode_value = lambda value: Binary(codec[0](value))
        self.decode_value = lambda value: codec[1](binary_type(value))
        if default is not None:
            self.update(default)

    def __setitem__(self, key, value):
        ''' Insert/update a key (uses upsert)

        ``key`` and ``value`` must be unicode or UTF-8.
        '''
        value = self.encode_value(value)
        return self._collection.update({'_id': key},
                                       {'_id': key, 'v': value},
                                       upsert=True)

    def __getitem__(self, key):
        ''' Return the value for key ``key``

        ``key`` must be unicode or UTF-8.
        If not found, raises ``KeyError``.
        '''
        result = self._collection.find({'_id': key},
                                       {'v': 1, '_id': 0})\
                                 .hint(self._index)
        if result.count() == 0:
            raise KeyError(key)
        return self.decode_value(result[0]['v'])

    def __delitem__(self, key):
        ''' Delete the key/value for key ``key``

        ``key`` must be unicode or UTF-8.
        If not found, raises ``KeyError``.
        '''
        if key not in self:
            raise KeyError(key)
        return self._collection.remove({'_id': key})

    def clear(self):
        ''' Delete all key/value pairs '''
        self._collection.remove({}, safe=self._safe)

    def __len__(self):
        ''' Return how many key/value pairs are stored '''
        return self._collection.count()

    def __iter__(self):
        ''' Iterate over all stored keys '''
        return (pair['_id']
                for pair in self._collection.find({}, {'_id': 1}))

    def __contains__(self, key):
        ''' Return True/False if a key is/is not stored in the collection '''
        results = self._collection.find({'_id': key}, {'_id': 1})
        return results.count() > 0

    has_key = __contains__

    def __del__(self):
        ''' Sync all operations and disconnect '''
        self._connection.fsync()
        self._connection.disconnect()
